routers/api/v1/admin/user.go

```diff
@@ -229,7 +229,9 @@ func EditUser(ctx *context.APIContext) {
 		ctx.ContextUser.MustChangePassword = *form.MustChangePassword
 	}
 
-	ctx.ContextUser.LoginName = form.LoginName
+	if len(form.LoginName) != 0 {
+		ctx.ContextUser.LoginName = form.LoginName
+	}
```

Comment:
`LoginName` gets set from `parseAuthSource` too.

Reply:
I would remove the lines 232-234 and add the check to the (faulty) `if sourceID == 0` in `parseAuthSource`. Don't make it too beautiful because that part will be replaced with #28733 anyway.

Reply:
I did add the check in `parseAuthSource`, but we can't remove 232-234 otherwise users will only be able to update LoginName when `sourceId != 0` 
(See [failing test when 232-234 is removed](https://github.com/go-gitea/gitea/actions/runs/7586140002/job/20663616375?pr=28834#step:6:177))

Reply:
As I said, the `sourceId != 0` is just wrong. `0` is the valid id of the default database login. With the admin api it's currently not possible to set a user login source to "use database".

https://github.com/go-gitea/gitea/blob/5574968ecbc34908dfa17b28bfc79c3490eaa685/models/auth/source.go#L276-L284

I think that check is there because if you don't provide a value in the request, it defaults to 0. Therefore you can't distinguish between "not set" and "set to 0". Other fields in the `EditUserOption` form use pointers which can be tested for `nil` to have a "is set" check. As `LoginName` is optional too now, both fields should be pointers too. You may need to change other tests then too and `util.ToPointer` will be your friend. 👍

Reply:
Thanks, this helped.
The fix for this case became messy, so I'll convert this to a draft for now, I'll pick it up once https://github.com/go-gitea/gitea/pull/28733 is merged.

Reply:
Can you take a look at it now, please?